File: ahu_dashboard/lib/models/ahu_telemetry.dart
import 'package:json_annotation/json_annotation.dart';

part 'ahu_telemetry.g.dart';

/// Telemetry data from ESP32 AHU unit
/// Supports both SHT45 (original) and SEN66+SDP810 (combo) sensors
@JsonSerializable()
class AhuTelemetry {
  // Basic readings (both sensor types)
  final double? temp;        // Temperature in °C
  final double? hum;         // Humidity in %RH
  final bool m1;             // Motor-1 (filter) status
  final bool m2;             // Motor-2 (drain) status
  final bool run;            // System running state
  final bool cp;             // Compressor (cooling) status
  final bool heater;         // Heater (dehumidifier) status
  final bool fan;            // Fan status (on/off)
  @JsonKey(name: 'fanSpeed') final int fanSpeed;  // Fan speed: 0=OFF, 1=LOW, 2=MED, 3=HIGH
  final double tempSet;      // Temperature setpoint
  final double humSet;       // Humidity setpoint
  final int ts;              // Timestamp (millis)
  
  // Sensor type indicator
  @JsonKey(name: 'sensorType') final String? sensorType;  // "sht45" or "combo"
  
  // SEN66 Air Quality readings (combo sensors only)
  final int? aqi;            // Air Quality Index
  @JsonKey(name: 'pm1p0') final double? pm1p0;    // PM1.0 µg/m³
  @JsonKey(name: 'pm2p5') final double? pm2p5;    // PM2.5 µg/m³
  @JsonKey(name: 'pm4p0') final double? pm4p0;    // PM4.0 µg/m³
  @JsonKey(name: 'pm10p0') final double? pm10p0;  // PM10.0 µg/m³
  final double? voc;         // VOC Index
  final double? nox;         // NOx Index
  final int? co2;            // CO2 in ppm
  
  // SDP810 HEPA readings (combo sensors only)
  @JsonKey(name: 'diffPressure') final double? diffPressure;  // Differential pressure Pa
  @JsonKey(name: 'hepaStatus') final String? hepaStatus;      // HEPA filter status
  @JsonKey(name: 'hepaHealth') final int? hepaHealth;         // HEPA health percentage

  AhuTelemetry({
    this.temp,
    this.hum,
    required this.m1,
    required this.m2,
    required this.run,
    required this.cp,
    required this.heater,
    required this.fan,
    required this.fanSpeed,
    required this.tempSet,
    required this.humSet,
    required this.ts,
    this.sensorType,
    this.aqi,
    this.pm1p0,
    this.pm2p5,
    this.pm4p0,
    this.pm10p0,
    this.voc,
    this.nox,
    this.co2,
    this.diffPressure,
    this.hepaStatus,
    this.hepaHealth,
  });

  factory AhuTelemetry.fromJson(Map<String, dynamic> json) =>
      _$AhuTelemetryFromJson(json);

  Map<String, dynamic> toJson() => _$AhuTelemetryToJson(this);

  /// Check if this is combo sensor data
  bool get isComboSensor => sensorType == 'combo';
  
  /// Check if air quality data is available
  bool get hasAirQualityData => aqi != null || pm2p5 != null || co2 != null;
  
  /// Check if HEPA data is available
  bool get hasHepaData => diffPressure != null || hepaStatus != null;

  /// HEPA thresholds by fan speed (Pa) - matches ESP32 logic
  /// Index: 0=OFF, 1=LOW, 2=MED, 3=HIGH
  static const List<double> _hepaMinNormal = [0.0, 40.0, 60.0, 90.0];
  static const List<double> _hepaMaxNormal = [0.0, 55.0, 90.0, 110.0];
  static const List<double> _hepaReplace = [0.0, 75.0, 120.0, 140.0];

  /// Calculate HEPA health percentage based on pressure and fan speed
  /// Returns 100% when in normal range, decreasing in clogging range
  int get calculatedHepaHealth {
    if (diffPressure == null) return hepaHealth ?? 0;
    
    final speedIdx = fanSpeed.clamp(0, 3);
    
    // When fan is OFF, show 0%
    if (speedIdx == 0) return 0;
    
    final absP = diffPressure!.abs();
    final minNormal = _hepaMinNormal[speedIdx];
    final maxNormal = _hepaMaxNormal[speedIdx];
    final replaceThreshold = _hepaReplace[speedIdx];
    
    if (absP < minNormal) {
      // Weak Airflow/Leak
      return 0;
    } else if (absP <= maxNormal) {
      // Normal - 100% health
      return 100;
    } else if (absP <= replaceThreshold) {
      // Clogging - decreasing health
      final health = 100.0 * (replaceThreshold - absP) / (replaceThreshold - maxNormal);
      return health.clamp(0, 100).toInt();
    } else {
      // Replace Required
      return 0;
    }
  }

  /// Get HEPA status based on pressure and fan speed
  String get calculatedHepaStatus {
    if (diffPressure == null) return hepaStatus ?? 'Unknown';
    
    final speedIdx = fanSpeed.clamp(0, 3);
    
    // When fan is OFF
    if (speedIdx == 0) return 'Fan Off';
    
    final absP = diffPressure!.abs();
    final minNormal = _hepaMinNormal[speedIdx];
    final maxNormal = _hepaMaxNormal[speedIdx];
    final replaceThreshold = _hepaReplace[speedIdx];
    
    if (absP < minNormal) {
      return 'Weak Airflow/Leak';
    } else if (absP <= maxNormal) {
      return 'Normal';
    } else if (absP <= replaceThreshold) {
      return 'Clogging';
    } else {
      return 'Replace Required';
    }
  }

  /// Get temperature display string
  String get tempDisplay => temp != null ? '${temp!.toStringAsFixed(1)}°C' : 'N/A';

  /// Get humidity display string
  String get humDisplay => hum != null ? '${hum!.toStringAsFixed(1)}%' : 'N/A';

  /// Check if sensor data is valid
  bool get hasSensorData => temp != null && hum != null;

  /// Get fan speed display string
  String get fanSpeedDisplay {
    switch (fanSpeed) {
      case 0:
        return 'OFF';
      case 1:
        return 'LOW (5V)';
      case 2:
        return 'MED (7V)';
      case 3:
        return 'HIGH (9V)';
      default:
        return 'UNKNOWN';
    }
  }
  
  /// Get AQI category
  String get aqiCategory {
    if (aqi == null) return 'N/A';
    if (aqi! <= 50) return 'Good';
    if (aqi! <= 100) return 'Moderate';
    if (aqi! <= 150) return 'Unhealthy (Sensitive)';
    if (aqi! <= 200) return 'Unhealthy';
    if (aqi! <= 300) return 'Very Unhealthy';
    return 'Hazardous';
  }
  
  /// Get AQI color
  int get aqiColorValue {
    if (aqi == null) return 0xFF9E9E9E;
    if (aqi! <= 50) return 0xFF4CAF50;    // Green
    if (aqi! <= 100) return 0xFFFFEB3B;   // Yellow
    if (aqi! <= 150) return 0xFFFF9800;   // Orange
    if (aqi! <= 200) return 0xFFF44336;   // Red
    if (aqi! <= 300) return 0xFF9C27B0;   // Purple
    return 0xFF880E4F;                     // Maroon
  }
  
  /// Get CO2 level description
  String get co2Level {
    if (co2 == null) return 'N/A';
    if (co2! <= 600) return 'Excellent';
    if (co2! <= 800) return 'Good';
    if (co2! <= 1000) return 'Moderate';
    if (co2! <= 1500) return 'Poor';
    return 'Ventilate!';
  }
}

File: pi_dashboard/app/models/ahu_telemetry.py
from dataclasses import dataclass, field
from typing import Optional


# HEPA differential-pressure thresholds per fan speed (Pa)
_HEPA_NORMAL = {1: (40, 55), 2: (60, 90), 3: (90, 110)}
_HEPA_REPLACE = {1: 75, 2: 120, 3: 140}

_AQI_CATEGORIES = [
    (0,   50,  "Good",            "#10B981"),
    (51,  100, "Moderate",        "#F59E0B"),
    (101, 150, "Unhealthy (Sens)","#F97316"),
    (151, 200, "Unhealthy",       "#EF4444"),
    (201, 300, "Very Unhealthy",  "#8B5CF6"),
    (301, 500, "Hazardous",       "#7F1D1D"),
]


@dataclass
class AhuTelemetry:
    temp: float = 0.0
    hum: float = 0.0
    m1: bool = False
    m2: bool = False
    run: bool = False
    cp: bool = False
    heater: bool = False
    fan: bool = False
    fan_speed: int = 0          # 0-3
    temp_set: float = 24.0
    hum_set: float = 55.0
    ts: int = 0

    # Optional combo-sensor fields
    sensor_type: Optional[str] = None
    aqi: Optional[float] = None
    pm1p0: Optional[float] = None
    pm2p5: Optional[float] = None
    pm4p0: Optional[float] = None
    pm10p0: Optional[float] = None
    voc: Optional[float] = None
    nox: Optional[float] = None
    co2: Optional[float] = None
    diff_pressure: Optional[float] = None
    hepa_status: Optional[str] = None
    hepa_health: Optional[float] = None

    @property
    def is_combo_sensor(self) -> bool:
        return self.sensor_type is not None

    @property
    def has_air_quality_data(self) -> bool:
        return self.pm2p5 is not None or self.co2 is not None

    @property
    def has_hepa_data(self) -> bool:
        return self.diff_pressure is not None

    @property
    def fan_speed_display(self) -> str:
        return ["OFF", "LOW", "MED", "HIGH"][self.fan_speed] if 0 <= self.fan_speed <= 3 else "?"

    @property
    def calculated_hepa_health(self) -> float:
        if self.diff_pressure is None or self.fan_speed == 0:
            return 100.0
        normal_range = _HEPA_NORMAL.get(self.fan_speed, (60, 90))
        replace_thresh = _HEPA_REPLACE.get(self.fan_speed, 120)
        dp = self.diff_pressure
        mid = (normal_range[0] + normal_range[1]) / 2
        if dp <= mid:
            return 100.0
        elif dp >= replace_thresh:
            return 0.0
        else:
            health = 1.0 - (dp - mid) / (replace_thresh - mid)
            return max(0.0, min(100.0, health * 100))

    @property
    def calculated_hepa_status(self) -> str:
        if self.diff_pressure is None or self.fan_speed == 0:
            return "Fan Off"
        h = self.calculated_hepa_health
        if h > 70:
            return "Normal"
        elif h > 40:
            return "Clogging"
        elif h > 10:
            return "Replace Required"
        else:
            return "Weak Airflow/Leak"

    @property
    def aqi_category(self) -> tuple:
        if self.aqi is None:
            return ("Unknown", "#94A3B8")
        v = int(self.aqi)
        for lo, hi, label, color in _AQI_CATEGORIES:
            if lo <= v <= hi:
                return (label, color)
        return ("Hazardous", "#7F1D1D")

    @property
    def co2_level(self) -> str:
        if self.co2 is None:
            return "N/A"
        v = self.co2
        if v < 400:   return "Fresh"
        if v < 1000:  return "Good"
        if v < 2000:  return "Moderate"
        return "High"

    @classmethod
    def from_dict(cls, d: dict) -> "AhuTelemetry":
        def g(key, alt=None):
            return d.get(key, alt)

        fan_speed = g("fanSpeed", g("fan_speed", 0))
        try:
            fan_speed = int(fan_speed)
        except (TypeError, ValueError):
            fan_speed = 0

        return cls(
            temp=float(g("temp", 0)),
            hum=float(g("hum", 0)),
            m1=bool(g("m1", False)),
            m2=bool(g("m2", False)),
            run=bool(g("run", False)),
            cp=bool(g("cp", False)),
            heater=bool(g("heater", False)),
            fan=bool(g("fan", False)),
            fan_speed=fan_speed,
            temp_set=float(g("tempSet", g("temp_set", 24.0))),
            hum_set=float(g("humSet", g("hum_set", 55.0))),
            ts=int(g("ts", 0)),
            sensor_type=g("sensorType") or g("sensor_type"),
            aqi=_float(g("aqi")),
            pm1p0=_float(g("pm1p0")),
            pm2p5=_float(g("pm2p5")),
            pm4p0=_float(g("pm4p0")),
            pm10p0=_float(g("pm10p0")),
            voc=_float(g("voc")),
            nox=_float(g("nox")),
            co2=_float(g("co2")),
            diff_pressure=_float(g("diffPressure") or g("diff_pressure")),
            hepa_status=g("hepaStatus") or g("hepa_status"),
            hepa_health=_float(g("hepaHealth") or g("hepa_health")),
        )


def _float(v) -> Optional[float]:
    try:
        return float(v) if v is not None else None
    except (TypeError, ValueError):
        return None
